# Deepssv: detecting somatic small variants in paired tumor and normal sequencing data with convolutional neural network

Deepssv takes as input a mixed pileup file generated by samtools from tumor and normal BAM files. It first operates on each genomic site independently to identify candidate somatic sites based on the criteria we have defineds. Next it encodes the mapping information samples that are readily available in the pileup format file around the candidate somatic sites into an array. Each array is a spatial representation of mapping information adapted for convolutional architecture. Then the convolutional neural network (CNN) model trained on experimentally validated somatic events evaluates the information in these arrays to obtain additional support for true positives and filter false positive predictions. Finally, potential somatic small variants determined by the CNN model are generated in the variant call format (VCF). 

Deepssv was tested on ubuntu 16.04 LTS and requires Python 3.5.2.

Prerequisites:

TensorFlow 1.8.0

Please see https://www.tensorflow.org/install/install_linux for how to install TensorFlow.







Please help us improve Deepssv, by reporting bugs or any ideas on how to make things better. You can submit an issue or send me an email.

Jing Meng        

jing.mengrabbit@gmail.com
